src/charm.py

```diff
@@ -26,38 +24,40 @@ class GitLabK8sCharm(CharmBase):
     def __init__(self, framework, key):
         super().__init__(framework, key)
 
-        for event in (self.on.start,
-                      self.on.upgrade_charm,
-                      self.on.config_changed,
-                      self.on.mysql_relation_changed):
-            self.framework.observe(event, self.on_start)
-        self.framework.observe(self.on.website_relation_joined, self)
-
-        self.website = HTTPInterfaceProvides(self, 'website')
-        self.mysql = MySQLInterfaceRequires(self, 'mysql')
+        self.website = HTTPServer(self, 'website')
+        self.mysql = MySQLServer(self, 'mysql')
         self.gitlab_image = OCIImageResource(self, 'gitlab_image')
 
+        self.state.is_started = getattr(self.state, 'is_started', False)
```

Comment:
I'm ok with `set_default_value` (or `set_default`; dicts use `setdefault` but the usual pattern in Python is to use underscores rather than camel case).

You could also use keyword style args, like `self.state.set_default(is_started=False)` which seems to me to give a little bit more of an impression of assignment (though admittedly not too much), but I'd also be on board with something like:

```python
class MyCharm(CharmBase):
    state = StoredState(
        is_started=False,
        has_website=False,
    )
```

Which keeps the defaults close to the declaration and seems to imply (to me at least) a bit more that they will be set as attributes.

Going even further down road, since we already have the `MyCharmEvents` pattern, we could also use subclassing for the stored state:

```python
class MyCharmState(StoredState):
    is_started = False
    has_website = False

class MyCharm(CharmBase):
    state = MyCharmState()
```

Reply:
I also want to call out that @niemeyer made a comment in passing during the call with Tom of calling the state variable `stored` instead of `state` to make it more clear that it was persisted.  I'm ambivalent about it, but we should be consistent; so far we've always used `state` that I've seen, so if we want to switch, we should do so everywhere ASAP and get used to that new name.

Reply:
```
self.stored.is_started = False
```

I could get used to it, but the fact that it is 'state = StoredState' already gives that hint, doesn't it?

I like the keyword style args, a lot more than quoting.

`__init__` seems ok, though it doesn't call out that these values will be ignored if they already exist. As a decorator, I also think the actual code might be a bit convoluted. (I think you have to set aside some sort of _default cache, and then grab from there if the value wasn't loaded.)

So my personal favorite is probably: `self.state.set_default(is_started=False)`